gflownet/envs/base.py

```diff
@@ -113,6 +136,46 @@ def actions2indices(
         # the action_dim dimension are True
         return torch.where(torch.all(actions == action_space, dim=2))[1]
 
+    def _get_state(self, state: Union[List, TensorType["state_dims"]]):
+        """
+        A helper method for other methods to determine whether state should be taken
+        from the arguments or from the instance (self.state): if is None, it is taken
+        from the instance.
+
+        Args
+        ----
+        state : list or tensor or None
+            None, or a state in GFlowNet format.
+
+        Returns
+        -------
+        state : list or tensor
+            The argument state, or self.state if state is None.
+        """
+        if state is None:
+            state = copy(self.state)
```

Comment:
It's a good question. I am sure it is not necessary to copy in all cases but I am also sure that it is in some cases and also that not copying is the source of a lot of silent bugs I have seen. Copying is safer and easier so I'd go with it for now, but I'm open to discuss an alternative, even though it may involve careful review of many parts of the code.

Reply:
The idea is to, in an upcoming update, use `_get_state()` in all the environments, instead of `if state is None...`. Perhaps that will be the chance to rethink whether we should copy or not or optionally.